### PR TITLE
Geom: Polygon and Rectangle

### PR DESCRIPTION
**Pull request by scveloso**
### Description
Continued work on new geom shape hierarchy by adding Polygon and Rectangle as well as tests for both.

- Fleshed out more of the new shape hierarchy for new geom
- Now has everything but Triangle

### Testing Done
Unit tests in ```new_geom/rectangle_test.cpp``` and ```new_geom/polygon_test.cpp```

### Resolved Issues

Partially addresses #958 

### Length Justification
Copying over ```segment``` to ```new_geom``` as well as the unit tests is why this is longer than 500 lines. It's not that much longer and most of the logic is fairly similar to that in old geom. 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/new_geom/convex_polygon.h [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#include "software/new_geom/convex_shape.h"
+#include "software/new_geom/polygon.h"
+
+/**
+ * A polygon that is convex (curved outwards).
+ */
+class ConvexPolygon : public virtual Polygon, public ConvexShape
```
Comment (jsn0716) [inline src/software/new_geom/convex_polygon.h]: What's the reason for virtual inheritance for `Polygon` here? Also, it seems like there are no guarantees being made that a `ConvexPolygon` is actually `convex`, and the `area` method doesn't seem to be implemented for any generic `ConvexPolygon`.

**Review by jsn0716 — CHANGES_REQUESTED** [review]
Looking good 👍

File: src/software/new_geom/shape.h [hunk 2]
```diff
@@ -14,11 +16,4 @@ class Shape
      * @return whether the Point p is contained within this Shape.
      */
     virtual bool contains(const Point &p) const = 0;
-
-    /**
-     * Returns the area of this Shape.
-     *
-     * @return The area of this Shape.
-     */
-    virtual double area() const = 0;
```
[the remark this answers is not in the document]
Comment (scveloso) [reply src/software/new_geom/shape.h:23]: There was no need for area in Polygon, only base classes like Circle and Rectangle had area implementation and uses so there wasnt any common behavior i can extract to the parent classes which is why I removed it.

File: src/software/new_geom/convex_polygon.h (hunk 1 above)
Comment (scveloso) [reply src/software/new_geom/convex_polygon.h]: Area method does not need to be implemented in ConvexPolygon, it is not virtual in any of the parent classes.

File: src/software/new_geom/shape.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsn0716) [reply src/software/new_geom/shape.h:23]: A general polygon can be defined as just a list of points, and it's a valid question to ask what the area of any such general polygon is that isn't necessarily a rectangle, triangle, etc. For example, this concave polygon has an area that can be calculated: https://en.wikipedia.org/wiki/Concave_polygon#/media/File:Simple_polygon.svg

File: src/software/new_geom/convex_polygon.h (hunk 1 above)
Comment (jsn0716) [reply src/software/new_geom/convex_polygon.h]: I should be able to create a generic `ConvexPolygon` and calculate its area. As is, I believe I can create any polygon using `ConvexPolygon` class without any checks on if it's actually a convex polygon, and also wouldn't be able to calculate its area. The `area` method also seems to already be written as `virtual` in `ConvexShape`.

File: src/software/new_geom/convex_shape.h [hunk 3]
```diff
@@ -3,6 +3,12 @@
 /**
  * A shape that is convex (curved outwards).
  */
-class ConvexShape : public Shape
+class ConvexShape : public virtual Shape
```
Comment (MathewMacDougall) [inline src/software/new_geom/convex_shape.h:6]: I don't think you need the `virtual` when inheriting like this

File: src/software/new_geom/rectangle.h [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    /**
+     * Returns a digit corresponding to a specific corner for pos%4. (0 for <-x,-y>, 1
+     * for <-x,+y>, 2 for <+x,+y> and 3 for <+x,-y>)
+     *
+     * @param pos The position of the corner wanted
+     *
+     * @return Point A digit corresponding to position
+     */
+    Point operator[](unsigned int pos) const;
```
Comment (MathewMacDougall) [inline src/software/new_geom/rectangle.h]: Do we actually use this function in our code? IMO I don't think we should have it since it exposes the internal representation of the `Rectangle`.

File: src/software/new_geom/rectangle.h [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    /**
+     * Returns the corner point of the rectangle that is the furthest from the input
+     * point.
+     *
+     * @param p The point to test
+     * @return The corner point that is furthest from the test point
+     */
+    Point furthestCorner(const Point &p);
```
Comment (MathewMacDougall) [inline src/software/new_geom/rectangle.h:85]: I don't recall this function from the previous geom Rectangle code. Is is used anywhere?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Looking good, just a few small comments

File: src/software/new_geom/polygon.h [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    const std::vector<Segment>& getSegments() const;
+
+    /**
+     * Returns the points that form the polygon.
+     * @return the points that form the polygon.
+     */
+    const std::vector<Point>& getPoints() const;
+
+   protected:
+    std::vector<Segment> segments_;
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/new_geom/polygon.h]: Agreed, I also think points are a better representation because it's not possible to construct a non-closed shape with them. Totally fine to have a `getSegments` method still.

File: src/software/new_geom/shape.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/new_geom/shape.h:23]: If it's trivial to do, then we can add it, otherwise I'd rather not add things until we have a known use-case for them (_IIRC, I think this is fairly non-trivial for arbitrary polygons, consider the case where the polygon is self-intersecting_)

File: src/software/new_geom/convex_shape.h (hunk 3 above)
Comment (garethellis0) [reply src/software/new_geom/convex_shape.h:6]: For context, checkout "The solution" on this page for an example of why you might want to do virtual inheritance like this: https://en.wikipedia.org/wiki/Virtual_inheritance

File: src/software/new_geom/convex_polygon.h (hunk 1 above)
Comment (garethellis0) [reply src/software/new_geom/convex_polygon.h]: - I don't think we need virtual inheritance here (see my reply to Mat's comment below for when it's appropriate)
- implementing `area()` shouldn't be too bad for convex polygons (see [here](http://www.mathwords.com/a/area_convex_polygon.htm)). It's definitely declared in `ConvexShape`, so it should definitely be implemented here. Then we can also remove it from the subclasses.
- Constructors aren't inherited, so don't think this allows creation of an invalid `ConvexPolygon`

File: src/software/new_geom/polygon.cpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    // A quick description of the algorithm (also from the same post) is as follows:
+    // "I run a semi-infinite ray horizontally (increasing x, fixed y) out from the test
+    // point, and count how many edges it crosses. At each crossing, the ray switches
+    // between inside and outside. This is called the Jordan curve theorem."
+    //
+    // NOTE: This algorithm will treat boundaries on the bottom-left of the polygon
+    // different from the boundaries on the top-right of the polygon. This small
+    // inconsistency does not matter for our use cases, and actually has the benefit that
+    // should two distinct polygons share an edge, any point along this edge will be
+    // located in one and only one polygon.
```
Comment (garethellis0) [inline src/software/new_geom/polygon.cpp:23]: Do we have tests that enforce this behavior with respect to boundaries on the bottom-left vs top-right?

File: src/software/new_geom/polygon_test.cpp [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    EXPECT_FALSE(hexagon.contains(Point(2.01, 0)));
+    EXPECT_FALSE(hexagon.contains(Point(-2.01, 0)));
+    EXPECT_FALSE(hexagon.contains(
+        Point(2.0f, 0.0f)));  // on right edge, see NOTE on Polygon::contains
+    EXPECT_TRUE(hexagon.contains(Point(-2, 0)));  // on left edge
+    EXPECT_TRUE(
+        hexagon.contains(Point(-2, -1)));  // the bottom left vertex of the hexagon
+    EXPECT_TRUE(hexagon.contains(Point(1, -1)));
+    EXPECT_TRUE(hexagon.contains(Point(-1.5, 0.75)));
+}
```
Comment (garethellis0) [inline src/software/new_geom/polygon_test.cpp:91]: Could we add a couple tests for self-intersecting polygons?

File: src/software/new_geom/rectangle.cpp [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+Point Rectangle::posXPosYCorner() const
+{
+    return negXNegYCorner() + diagonal();
+}
+
+Point Rectangle::negXPosYCorner() const
+{
+    return negXNegYCorner() + Vector(0, diagonal().y());
+}
```
Comment (garethellis0) [inline src/software/new_geom/rectangle.cpp:43]: For each of the "corner" functions here, could we just return a specific point from `points_` since we enforce the order in the constructor?

File: src/software/new_geom/rectangle.h [hunk 10]
```diff
@@ -0,0 +1,112 @@
+#include "software/new_geom/convex_polygon.h"
+
+/**
+ * A rectangle.
```
Comment (garethellis0) [inline src/software/new_geom/rectangle.h]: Could we indicate that this rectangle cannot be rotated?

File: src/software/new_geom/segment.h [hunk 11]
```diff
@@ -0,0 +1,88 @@
+#pragma once
+
+#include "software/new_geom/point.h"
+
+// This is a direct copy of geom/segment.h, this is needed to unblock work on the new_geom
+// shape hierarchy.
+// TODO: Refactor this to fit the new_geom hierarchy
```
Comment (garethellis0) [inline src/software/new_geom/segment.h]: Is there a ticket for this? If not, let's file on, and either way, we should note the ticket here (ex. `TODO: Issue #1234`)

**Review by garethellis0 — CHANGES_REQUESTED** [review]
This is looking good, super happy to see this.

File: src/software/new_geom/convex_polygon.h (hunk 1 above)
Comment (scveloso) [reply src/software/new_geom/convex_polygon.h]: I have virtual inheritance here because ```Rectangle``` uses ```Polygon```'s constructor and cannot do so unless ```Polygon``` is either a virtual or direct base of ```Rectangle```. Is there an alternative both of you were thinking of here?

Comment (garethellis0) [reply src/software/new_geom/convex_polygon.h]: Ah, I see. That's all good then.

File: src/software/new_geom/rectangle.h (hunk 5 above)
Comment (scveloso) [reply src/software/new_geom/rectangle.h:85]: Its used in ```voronoi_diagram.cpp```, and old geom ```util.cpp``` - its declared in old geom ```rectangle.cpp```

File: src/software/new_geom/polygon.cpp (hunk 7 above)
Comment (scveloso) [reply src/software/new_geom/polygon.cpp:23]: We do, should be in ```polygon_test.cpp```

File: src/software/new_geom/shape.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/new_geom/shape.h:23]: Might be better to implement in `ConvexPolygon` then since you can guarantee the polygon is non-intersecting.

File: src/software/new_geom/polygon.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (scveloso) [reply src/software/new_geom/polygon.h]: Addressed in latest update

File: src/software/new_geom/rectangle.h (hunk 4 above)
Comment (scveloso) [reply src/software/new_geom/rectangle.h]: Removed in latest update

File: src/software/new_geom/polygon.cpp [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    bool point_is_contained = false;
+    unsigned i              = 0;
+    unsigned j              = points_.size() - 1;
+    while (i < points_.size())
+    {
+        if (((points_.at(i).y() > p.y()) != (points_.at(j).y() > p.y())) &&
+            (p.x() < (points_.at(j).x() - points_.at(i).x()) *
+                             (p.y() - points_.at(i).y()) /
+                             (points_.at(j).y() - points_.at(i).y()) +
+                         points_.at(i).x()))
```
[the remark this answers is not in the document]
Comment (scveloso) [reply src/software/new_geom/polygon.cpp]: Fixed in latest update

File: src/software/new_geom/polygon.h [hunk 13]
```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include <vector>
+
+#include "software/new_geom/segment.h"
+#include "software/new_geom/shape.h"
+
+/**
+ * A shape composed of segments.
```
[the remark this answers is not in the document]
Comment (scveloso) [reply src/software/new_geom/polygon.h]: Fixed in latest update

File: src/software/new_geom/polygon.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (scveloso) [reply src/software/new_geom/polygon.cpp:23]: Added tests for self-intersecting polygons in latest update

File: src/software/new_geom/polygon_test.cpp (hunk 8 above)
Comment (scveloso) [reply src/software/new_geom/polygon_test.cpp:91]: Done in latest update

File: src/software/new_geom/rectangle.cpp (hunk 9 above)
Comment (scveloso) [reply src/software/new_geom/rectangle.cpp:43]: Fixed in latest update

File: src/software/new_geom/rectangle.h (hunk 10 above)
Comment (scveloso) [reply src/software/new_geom/rectangle.h]: Fixed in latest update

File: src/software/new_geom/segment.h (hunk 11 above)
Comment (scveloso) [reply src/software/new_geom/segment.h]: Made issue #1098 and referenced it in the TODO, done in latest update

**Review by MathewMacDougall — DISMISSED** [review]

File: src/software/new_geom/BUILD [hunk 14]
```diff
@@ -5,18 +5,24 @@ cc_library(
...
         "convex_shape.h",
         "geom_constants.h",
         "point.h",
+        "polygon.h",
         "polynomial.h",
         "ray.h",
+        "rectangle.h",
+        "segment.h",
         "shape.h",
         "vector.h",
```
Comment (garethellis0) [inline src/software/new_geom/BUILD]: @jsn0716 - just a heads up on the merge conflicts this will probably cause with your PR.

File: src/software/new_geom/polygon.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/new_geom/polygon.cpp]: Definitely better now, still having a bit of trouble reading the following:
```
  (p.x() < (points_.at(j).x() - points_.at(i).x()) *
                             (p.y() - points_.at(i).y()) /
                             (points_.at(j).y() - points_.at(i).y()) +
                         points_.at(i).x());
```
Is there any way we can break out some parts of this expression into values? No worries if not, just want to make sure we use all our options to make this as understandable as possible.

File: src/software/new_geom/polygon_test.cpp [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+        Point(-1, -1)));  // on a "top" edge, see NOTE on Polygon::contains
+    EXPECT_FALSE(intersecting_poly.contains(
+        Point(0, -1.5)));  // on a "right" edge, see NOTE on Polygon::contains
+    EXPECT_TRUE(intersecting_poly.contains(Point(-1, -2)));
+    EXPECT_TRUE(intersecting_poly.contains(Point(0, 1)));
+    EXPECT_TRUE(intersecting_poly.contains(Point(1, 1.5)));
+    EXPECT_TRUE(intersecting_poly.contains(Point(1, 1)));
+    EXPECT_TRUE(intersecting_poly.contains(Point(0, 1.5)));
+    EXPECT_TRUE(intersecting_poly.contains(Point(-1, -1.5)));
+}
```
Comment (garethellis0) [inline src/software/new_geom/polygon_test.cpp:172]: One last test to suggest:
```
XXXXXXXXXXXXXXXXXXXXXXXXXXXX
X                          XX
X                           X
X         XXXXXXXX          X
X         X      X          X
X          XX   X           X
X            XXXXX          X
X          XXX   XXXXX     XX
XXXXXXXXXXXX         XXXXXXX
```
(apologies for the horrid ascii drawing, just wanted to quickly convey the idea).

The impetus for suggesting this is I just want to be sure we're being explicit what's "contained" in a polygon like this.

**Comment by garethellis0** [comment]
Just one additional test and then I think all my comments are basically resolved!

File: src/software/new_geom/convex_polygon.h (hunk 1 above)
Comment (jsn0716) [reply src/software/new_geom/convex_polygon.h]: Could we implement the `area()` function for `ConvexPolygon` @scveloso?

**Comment by jsn0716** [comment]
@scveloso I broke up the libraries for `new_geom/BUILD` in #1094, looks like it caused a merge conflict here

File: src/software/new_geom/polygon_test.cpp (hunk 15 above)
Comment (scveloso) [reply src/software/new_geom/polygon_test.cpp:172]: Added a test for this. Such loops are treated as "holes" and any points in them aren't contained.

File: src/software/new_geom/polygon.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (scveloso) [reply src/software/new_geom/polygon.cpp]: Tried to look into it more https://wrf.ecse.rpi.edu//Research/Short_Notes/pnpoly.html
But the stackoverflow and the link above aren't very clear on the conditions. Feel free to look into it as well maybe you can decipher something from it

File: src/software/new_geom/convex_polygon.h (hunk 1 above)
Comment (scveloso) [reply src/software/new_geom/convex_polygon.h]: Added ```area()``` for ```ConvexPolygon``` in latest update. Also removed ```area``` from ```Rectangle``` and made ```ConvexPolygon``` an abstract class.

File: src/software/new_geom/convex_polygon.cpp [hunk 16]
```diff
@@ -0,0 +1,29 @@
+#include "software/new_geom/convex_polygon.h"
+
+ConvexPolygon::ConvexPolygon(const std::initializer_list<Point>& points) : Polygon(points)
+{
+}
+
+double ConvexPolygon::area() const
```
Comment (MathewMacDougall) [inline src/software/new_geom/convex_polygon.cpp:45]: Can we add unit tests for this? Currently there a a few tests for rectangle area, but we want to test this for arbitrary Convex polygons

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/software/new_geom/convex_polygon.cpp (hunk 16 above)
Comment (scveloso) [reply src/software/new_geom/convex_polygon.cpp:45]: ConvexPolygon is an abstract class so we cant actually construct arbitrary polygons, this is meant to be used for rectangle, triangle, and any other convex polygons we implement in the future. Is this okay or did we wanna go another way with this?

Comment (MathewMacDougall) [reply src/software/new_geom/convex_polygon.cpp:45]: Ah I see. I think we would want to be able to create arbitrary (convex) polygons, since that is/was what our obstacle system used for navigation. I don't remember if we made an explicit decision about being able to construct arbitrary polygons in our big geom meeting a while ago.

@garethellis0 @jonathanlew @jsn0716 for opinions?

Comment (garethellis0) [reply src/software/new_geom/convex_polygon.cpp:45]: I don't recall the exact details of what we discussed in the original architecture meeting, but it's entirely reasonable for both `Polygon` and `ConvexPolygon` to be concrete classes. Certainly we can implement `area()` for the `ConvexPolygon` class, even if it is an abstract class. 

TL;DR: Both `Polygon` and `ConvexPolygon` should be concrete classes, `area` should be implemented for `ConvexPolygon`.

File: src/software/new_geom/polygon_test.cpp (hunk 15 above)
Comment (garethellis0) [reply src/software/new_geom/polygon_test.cpp:172]: Sorry if I'm just missing it, but I don't see this test here?

File: src/software/new_geom/polygon.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/new_geom/polygon.cpp]: In a bit of a rush now, but will try to  look at this tmw morning.

[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/new_geom/polygon.cpp]: I think all this is essentially doing vector interception. Take a look through https://stackoverflow.com/questions/217578/how-can-i-determine-whether-a-2d-point-is-within-a-polygon and let me know if that makes more sense. No worries if not, happy to spend some time breaking this out some more!

File: src/software/new_geom/polygon_test.cpp (hunk 15 above)
Comment (scveloso) [reply src/software/new_geom/polygon_test.cpp:172]: Check line 174 of polygon_test.cpp, something weird was going on - it wasnt committed in my previous commit for some reason. Should be there now.

File: src/software/new_geom/convex_polygon.cpp (hunk 16 above)
Comment (scveloso) [reply src/software/new_geom/convex_polygon.cpp:45]: ```ConvexPolygon``` is now a concrete class in the latest update

**Review by MathewMacDougall — DISMISSED** [review]
Nice work! The new concrete classes with convex and area stuff is super clean :+1:  LGTM pending Gareth's comments

File: src/software/new_geom/convex_polygon.h [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+     * Returns the area of this convex polygon
+     *
+     * @return The area of this convex polygon
+     */
+    double area() const override;
+
+    /**
+     * Construct a convex polygon by drawing line segments between consecutive
+     * Points, and then from the last Point to the first Point.
+     * @param points Points that form a convex polygon
```
Comment (garethellis0) [inline src/software/new_geom/convex_polygon.h:22]:
``` suggestion
     * 
     * @param points Points that form a convex polygon
     * 
     * @throws std::invalid_argument if the given points do not form a 
     *                               convex polygon
```

File: src/software/new_geom/convex_polygon.h [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+     */
+    explicit ConvexPolygon(const std::vector<Point>& points);
+
+    /**
+     * Construct a convex polygon by drawing line segments between consecutive
+     * Points, and then from the last Point to the first Point.
+     *
+     * @param points Points that form a convex polygon
```
Comment (garethellis0) [inline src/software/new_geom/convex_polygon.h:33]:
``` suggestion
     * @param points Points that form a convex polygon
     * 
     * @throws std::invalid_argument if the given points do not form a 
     *                               convex polygon
```

**Comment by garethellis0** [comment]
Just two suggestions (left above).

There's also one comment still open about simplifying some logic,  but I'll leave it up to your discretion if you see a reasonable way with what I linked. Happy to merge as-is if you don't see any easy simplifications.

**Comment by scveloso** [comment]
@garethellis0 Tried to look into that last comment of yours and the link you mentioned briefly mentions that it uses ray casting but makes no mention of the weird condition being used in the code. Feel free to have a go at it, maybe you can see something that I can't

**Comment by scveloso** [comment]
Also committed your suggestions to adding the @throws to ConvexPolygon constructors, all comments should be addressed now

**Review by MathewMacDougall — DISMISSED** [review]

File: src/software/new_geom/convex_polygon.cpp [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+
+        if (vertexAngleInDegrees > 180)
+        {
+            return false;
+        }
+
+        totalAngle += (180 - vertexAngleInDegrees);
+    }
+
+    return totalAngle == 360;
```
Comment (jsn0716) [inline src/software/new_geom/convex_polygon.cpp]: Since `totalAngle` is of type `double`, it should be compared using epsilon

**Review by jsn0716 — CHANGES_REQUESTED** [review]
Looking good, just one comment

File: src/software/new_geom/convex_polygon.cpp (hunk 19 above)
Comment (scveloso) [reply src/software/new_geom/convex_polygon.cpp]: Yeah you're right. Fixed this in the latest update.

**Review by MathewMacDougall — APPROVED** [review]

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Review by jsn0716 — APPROVED** [review]
LGTM 💯
